### PR TITLE
feat: containerize a2ln-server

### PR DESCRIPTION
**Pull request by yozachar**
- improves `.gitignore`
- source changes:
  - type ignore `gi.repository.Notify`
  - raise RuntimeError if `own_secret_key` is `None`
  - catch & clean exit on the same error
  - do not run sub process in a shell, if signals are to be passed.
- adds `containerfile`, `compose.yaml` & `.containerignore`

### CODE REVIEW

**Comment by joe733** [comment]
### This is the current output

```console
$ podman images
REPOSITORY                   TAG         IMAGE ID      CREATED         SIZE
localhost/a2ln               1.1.10      bf8f991db64d  11 minutes ago  623 MB

$ podman container ls
CONTAINER ID  IMAGE                  COMMAND          CREATED         STATUS         PORTS       NAMES
1346a6d87455  localhost/a2ln:1.1.10  /bin/sh -c a2ln  11 minutes ago  Up 11 minutes              a2ln

$ podman logs a2ln         
==> Notification server running on IP * and port 23045.
Do not forget to autostart the notification server. More information can be found at https://patri9ck.dev/a2ln/server.html#autostarting.
                             
                             
    █▀▀▀▀▀█ ▄▀█ ▀ █▀▀▀▀▀█    
    █ ███ █ █▄▀▄▄ █ ███ █    
    █ ▀▀▀ █  █▄▄▀ █ ▀▀▀ █    
    ▀▀▀▀▀▀▀ ▄█▀ ▀ ▀▀▀▀▀▀▀    
    ▄ ▀  ▄▀▄▀█▀▀▄██ ▄▀█▀     
    █▀▀▄ ▀▀██▀█ █▄▀▀▄▀██▄    
     ▀  ▀▀▀▀▄▄▄██▄▄▀▄▀▀▀▀    
    █▀▀▀▀▀█  ▄ ▀█▀█▀▄ █▄▄    
    █ ███ █  ▄█ █  █ ▄▀ ▀    
    █ ▀▀▀ █  █▄█▀█▀▀ █▀▀█    
    ▀▀▀▀▀▀▀ ▀▀   ▀  ▀ ▀   

$
```

For some reason, can't see logs from the pairing server.

@patri9ck, any idea why?

--- 

### How to test

1. Checkout [PR #32](https://stackoverflow.com/questions/27567846/how-can-i-check-out-a-github-pull-request-with-git#30584951).
2. Run the following commands, replace `podman` with `docker`.

```console
$ podman build . -f ./containerfile -t localhost/a2ln:1.1.10
$ podman run --rm --network host --name a2ln -t localhost/a2ln:1.1.10
```

Or if you've `podman-` or `docker-` compose.

```console
$ podman-compose -p a2ln-server -f ./compose.yaml up -d
```

**Comment by patri9ck** [comment]
Looks great! Will take a look at it the next few days and I will also try to fix that logs from the pairing server are missing.

File: src/a2ln/a2ln.py [hunk 1]
```diff
@@ -235,7 +240,7 @@ def replace(text: str) -> str:
                                      daemon=True).start()
 
                     if self.command is not None:
-                        subprocess.Popen(replace(self.command), shell=True)
+                        subprocess.Popen(replace(self.command), shell=False)
```
Comment (patri9ck) [inline src/a2ln/a2ln.py:243]: @yozachar Wouldn't  `shell=False` break some things?  https://docs.python.org/3/library/subprocess.html#frequently-used-arguments shows that `shell=True` allows you to use shell features such as pipes. Isn't this very useful?

Comment (yozachar) [reply src/a2ln/a2ln.py:243]: https://github.com/patri9ck/a2ln-server/blob/bfb2889257e510a7b38bad984e51e25d33707ba8/src/a2ln/a2ln.py#L119-L120

@patri9ck, command passed as values to `--command` gets executed as a sub-process, right?

When you set `shell=True` (and pass in, the user provided command, to shell), you're opening up your code to some potential security holes. Ref: https://docs.python.org/3/library/subprocess.html#security-considerations

A majority of  simple commands can be executed with `shell=False`, I recommend that you keep it so and when a specific and strong use-cases comes up, you can conditionally set `shell=True`.

Also if you do need to use shell pipes, you can do so:

```py
import subprocess
p1 = subprocess.Popen(["cat", "file.log"], stdout=subprocess.PIPE)
p2 = subprocess.Popen(["tail", "-1"], stdin=p1.stdout, stdout=subprocess.PIPE)
p1.stdout.close()  # Allow p1 to receive a SIGPIPE if p2 exits.
output,err = p2.communicate()
```

Ref: https://stackoverflow.com/a/7353141

**Comment by patri9ck** [comment]
@yozachar `podman logs a2ln ` works fine for me and also shows the logs of the pairing server.

But wouldn't we need to expose the ports of the pairing and notification server explicitly? If yes, then we would need to set a static port for the pairing server which is right now random by default.

**Comment by yozachar** [comment]
> But wouldn't we need to expose the ports of the pairing and notification server explicitly? If yes, then we would need to set a static port for the pairing server which is right now random by default.

Ah yes, which is why this PR was in draft. What changes to the PR do you suggest?

**Comment by patri9ck** [comment]
Is Docker actually able to access the libnotify instance of the host system?

**Comment by yozachar** [comment]
To test whether or not it does. I need to able to pair first:

```console
podman start -a a2ln
[a2ln-server] | ==> Notification server running on IP * and port 23045.
[a2ln-server] | Do not forget to autostart the notification server. More information can be found at https://patri9ck.dev/a2ln/server.html#autostarting.
[a2ln-server] |                              
[a2ln-server] |                              
[a2ln-server] |     █▀▀▀▀▀█ ▄▀ █  █▀▀▀▀▀█    
[a2ln-server] |     █ ███ █ █▄▀▄  █ ███ █    
[a2ln-server] |     █ ▀▀▀ █  ██▄▀ █ ▀▀▀ █    
[a2ln-server] |     ▀▀▀▀▀▀▀ █ ▀ ▀ ▀▀▀▀▀▀▀    
[a2ln-server] |     █▄▄ ▄ ▀ █▄▀▀▄██ ▄▀█▀     
[a2ln-server] |     ▄▀  ▄ ▀ ▄█▀ █▄▀▀▄▀██▄    
[a2ln-server] |      ▀▀   ▀ ███▀▀▄█▀▄▀▀▀▀    
[a2ln-server] |     █▀▀▀▀▀█  ▄█  ▄▀▀▄ █▄▄    
[a2ln-server] |     █ ███ █  ▀ ▀▄███ ▄▀ ▀    
[a2ln-server] |     █ ▀▀▀ █  ▀▄▀█▀█▀ █▀▀█    
[a2ln-server] |     ▀▀▀▀▀▀▀ ▀▀▀▀▀▀▀ ▀ ▀      
[a2ln-server] |                              
[a2ln-server] |                              
[a2ln-server] | ==> Pairing server running on IP * and port 38059.
[a2ln-server] | To pair a new device, open the Android 2 Linux Notifications app and scan this QR code or enter the following:
[a2ln-server] | IP: 192.168.0.15
[a2ln-server] | Port: 38059
[a2ln-server] | ==> Public Key: 7xYMPkuV[/C}UuzqP^yXaVQ+#zIgIxDWR>HcL}F=
[a2ln-server] | ==> New pairing request
[a2ln-server] | IP: 192.168.0.16
[a2ln-server] | Public Key: NqD.4Ihei#ZYBTy}@e1&!Iie4qE!jJx)s[0JtPSJ
Exception in thread Thread-3:
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/threading.py", line 1052, in _bootstrap_inner
    self.run()
  File "/home/a2ln/.local/lib/python3.12/site-packages/a2ln/a2ln.py", line 304, in run
    if input("Accept? (Yes/No): ").lower() != "yes":
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^
EOFError: EOF when reading a line
^CTraceback (most recent call last):
  File "/home/us-er/.local/bin/podman-compose", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/us-er/.local/pipx/venvs/podman-compose/lib/python3.11/site-packages/podman_compose.py", line 3205, in main
    podman_compose.run()
  File "/home/us-er/.local/pipx/venvs/podman-compose/lib/python3.11/site-packages/podman_compose.py", line 1524, in run
    retcode = cmd(self, args)
              ^^^^^^^^^^^^^^^
  File "/home/us-er/.local/pipx/venvs/podman-compose/lib/python3.11/site-packages/podman_compose.py", line 1920, in wrapped
    return func(*args, **kw)
           ^^^^^^^^^^^^^^^^^
  File "/home/us-er/.local/pipx/venvs/podman-compose/lib/python3.11/site-packages/podman_compose.py", line 2304, in compose_up
    thread.join(timeout=1.0)
  File "/usr/lib/python3.11/threading.py", line 1116, in join
    self._wait_for_tstate_lock(timeout=max(timeout, 0))
  File "/usr/lib/python3.11/threading.py", line 1132, in _wait_for_tstate_lock
    if lock.acquire(block, timeout):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
KeyboardInterrupt
```` 

Shall I make necessary changes to code that might fix this?

**Comment by patri9ck** [comment]
> To test whether or not it does. I need to able to pair first:
> 
> ```
> podman start -a a2ln
> [a2ln-server] | ==> Notification server running on IP * and port 23045.
> [a2ln-server] | Do not forget to autostart the notification server. More information can be found at https://patri9ck.dev/a2ln/server.html#autostarting.
> [a2ln-server] |                              
> [a2ln-server] |                              
> [a2ln-server] |     █▀▀▀▀▀█ ▄▀ █  █▀▀▀▀▀█    
> [a2ln-server] |     █ ███ █ █▄▀▄  █ ███ █    
> [a2ln-server] |     █ ▀▀▀ █  ██▄▀ █ ▀▀▀ █    
> [a2ln-server] |     ▀▀▀▀▀▀▀ █ ▀ ▀ ▀▀▀▀▀▀▀    
> [a2ln-server] |     █▄▄ ▄ ▀ █▄▀▀▄██ ▄▀█▀     
> [a2ln-server] |     ▄▀  ▄ ▀ ▄█▀ █▄▀▀▄▀██▄    
> [a2ln-server] |      ▀▀   ▀ ███▀▀▄█▀▄▀▀▀▀    
> [a2ln-server] |     █▀▀▀▀▀█  ▄█  ▄▀▀▄ █▄▄    
> [a2ln-server] |     █ ███ █  ▀ ▀▄███ ▄▀ ▀    
> [a2ln-server] |     █ ▀▀▀ █  ▀▄▀█▀█▀ █▀▀█    
> [a2ln-server] |     ▀▀▀▀▀▀▀ ▀▀▀▀▀▀▀ ▀ ▀      
> [a2ln-server] |                              
> [a2ln-server] |                              
> [a2ln-server] | ==> Pairing server running on IP * and port 38059.
> [a2ln-server] | To pair a new device, open the Android 2 Linux Notifications app and scan this QR code or enter the following:
> [a2ln-server] | IP: 192.168.0.15
> [a2ln-server] | Port: 38059
> [a2ln-server] | ==> Public Key: 7xYMPkuV[/C}UuzqP^yXaVQ+#zIgIxDWR>HcL}F=
> [a2ln-server] | ==> New pairing request
> [a2ln-server] | IP: 192.168.0.16
> [a2ln-server] | Public Key: NqD.4Ihei#ZYBTy}@e1&!Iie4qE!jJx)s[0JtPSJ
> Exception in thread Thread-3:
> Traceback (most recent call last):
>   File "/usr/local/lib/python3.12/threading.py", line 1052, in _bootstrap_inner
>     self.run()
>   File "/home/a2ln/.local/lib/python3.12/site-packages/a2ln/a2ln.py", line 304, in run
>     if input("Accept? (Yes/No): ").lower() != "yes":
>        ^^^^^^^^^^^^^^^^^^^^^^^^^^^
> EOFError: EOF when reading a line
> ^C
> ```
> 
> Shall I make necessary changes to code that might fix this?

If it isn't too complex, that would be great. But I'm almost certain that it won't work. For testing whether it is possible, it should be enough to just remove this check and to automatically accept the pairing request.

**Comment by yozachar** [comment]
Oh, no I just synced with `main`. I'll open a new PR.

**Pull request closed without merging**